### ISSUE TITLE
Wrong evaluation of injected service class when using trait

### ISSUE THREAD

**Issue by josefsabl** [nette/di#169]
Version: 2.4.13

### Bug Description
When property annotated with @inject exists in presenter and also in trait, and the class is denoted by relative name, the namespace of same property from trait is used.

### Steps To Reproduce
Let's say we have a trait we intend to use in a presenter that looks like this:

```php
namespace App;

trait TPresenter {
	/** @var MyService @inject */
	public $myService;
}
```

The service looks like this:

```php
namespace App;

class MyService {	
}
```

Now we use it in presenters like this (note the namespace).

```php
namespace App\FrontModule;

final class IndexPresenter extends \App\Presenter
{
	use \App\TPresenter;
	
	public function actionDefault()
	{
		bdump(get_class($this->myService)); // App\MyService
	}
}
```

This works just fine. But for some reason we want to use the custom service (again, notice the namespace) that looks like this in one of our presenters.

```php
namespace App\FrontModule;

class MyService {
}
```

So we alter the presenter code to look like this. Notice that although the annotation looks exactly same like the one in the trait, we actually are in a different namespace and the class should be `App\FrontModule\MyService`. But it still injects the `App\MyService`.

```php
namespace App\FrontModule;

final class IndexPresenter extends \App\Presenter
{
	use \App\TPresenter;
	
	/** @var MyService @inject */
	public $myService;
	
	public function actionDefault()
	{
		bdump(get_class($this->myService)); // Should be App\FrontModule\MyService, but is still App\MyService
	}
}
```

### Expected Behavior
The `App\FrontModule\MyService` should be injected, `App\MyService` is injected instead.

Another thing is that when the `App\FrontModule\MyService` is renamed to `App\FrontModule\MyService2` I get this error: `Class or interface 'App\MyService2' used in @var annotation at App\TPresenter::$myService not found.` This annotation is not in the `App\TPresenter` so I guess this is a bug in DI evaluation of class to inject?

### Possible Solution
Providing fully qualified class name fixes the problem but obviously is only a temporary and dangerous hack.

**Comment by dg** [comment]
I think it can not be solved. There is no way to find out that the property was (re)defined in the class.

**Comment by josefsabl** [comment]
Thanks for the comment. I traced the flow to `Nette/Utils/Reflection::getPropertyDeclaringClass` and I understand now. This bug in PHP is documented [here](https://bugs.php.net/bug.php?id=64963).

[This](http://mouf-php.com/blog/php_reflection_api_traits) is also an interesting post. I am not sure if it would work in Nette and it also feels pretty hackish.

So, lets hope the PHP guys will fix this anytime :-)

**Comment by josefsabl** [comment]
On second thought that "hack" by the David guy may work pretty well. He didn't succeeded because he needed to get much more info than what Nette needs. Also by reading it I realize that maybe the nested traits (traits using traits) might not be supported?

**Comment by milo** [comment]
That workaround will not work. Author assumes, that class cannot redeclare a property defined by trait, but it is no longer valid since PHP 7. Same properties are allowed when same visibility and default value is used.

IMHO, the only workaround is `token_get_all()`.

**Comment by milo** [comment]
And when I'm thinking about that, if redeclared property must match by visibility and by default value, it should match its type too (in the same way as method's return type). Ideal check task for static analyzer.

**Comment by JanTvrdik** [comment]
> IMHO, the only workaround is token_get_all().

I think it should be possible to solve this by analyzing `ReflectionClass::getTraits` and `ReflectionClass::getTraitAliases`. See the following code from PHPStan https://github.com/phpstan/phpstan/commit/7bde475e8d3ab8827ce18ec24e051b055a9e774d

**Comment by JanTvrdik** [comment]
@dg I'm may have a partial solution. `Nette\Utils\Reflection::getPropertyDeclaringClass()` may return the trait property reflection only when doc comment matches with class property reflection.

**Comment by milo** [comment]
I would not workaround it. When typed properties will be common, different types on trait and class will not be possible.

**Comment by dg** [comment]
@JanTvrdik I do not know if I understand you, but the above example describes the state where the doc comments are the same but have a different meaning.

**Comment by JanTvrdik** [comment]
@dg Yes, it will not work for the posted example (that's why I said it's a *partial* solution), but it will for some cases.

**Comment by dg** [comment]
It is true that now it is unreliable and it would be less unreliable.
